### PR TITLE
Runge neuralNetwork

### PR DESCRIPTION
**Pull request by emmyrunge**

### CODE REVIEW

File: src/main/java/smartsnake/neuralnetwork/runge/NeuralNetworkBrain.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    // return the result as a Direction
+    private final NeuralNetwork neuralNetwork;
+
+    public NeuralNetworkBrain() throws IOException {
+        neuralNetwork = NeuralNetwork.readFromFile();
+    }
+
+    @Override
+    public Direction move(Snake snake, Food food, Garden garden) {
+        NeuralNetworkDataFactory neuralNetworkDataFactory = new NeuralNetworkDataFactory();
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/runge/NeuralNetworkBrain.java:24]: create this only once.

File: src/main/java/smartsnake/neuralnetwork/runge/LearnAndSave.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                //and the move that your AstarBrain decides to make
+                neuralNetwork.train(neuralNetworkDataFactory.toInput(garden),
+                        neuralNetworkDataFactory.toOutput(snake.getDirection()));
+                if (!garden.tick())
+                {
+                    break;
+                }
+                //save the NN, found in NN class
+                //neuralNetwork.writeToFile();
+                neuralNetwork.writeToFile();
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/runge/LearnAndSave.java:41]: Write this once at the end, not every time through the loop.

File: src/main/java/smartsnake/neuralnetwork/runge/LearnAndSave.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            AstarBrain brain = new AstarBrain();
+            snake.setBrain(brain);
+
+            //while snake can move
+            while (true)
+            {
+
+                //for every move, train the NN with the state of the garden,
+                //and the move that your AstarBrain decides to make
+                neuralNetwork.train(neuralNetworkDataFactory.toInput(garden),
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/runge/LearnAndSave.java:33]: `toInput(garden)` should go before `garden.tick()` and `train()` should go afterwards.

Comment (andrewoid) [reply src/main/java/smartsnake/neuralnetwork/runge/LearnAndSave.java:32]: It needs to be 
1. toInput()
2. tick()
3. train()

You now have
1. toInput()
2. train()
3. tick()

File: src/main/java/smartsnake/neuralnetwork/runge/LearnAndSave.java [hunk 4]
```diff
@@ -25,21 +25,16 @@ public static void main(String[] args) {
...
-                    break;
-                }
-                //save the NN, found in NN class
-                //neuralNetwork.writeToFile();
-                neuralNetwork.writeToFile();
-            }
+            } while (garden.tick());
         }
+        //save the NN, found in NN class
+        //neuralNetwork.writeToFile();
```
Comment (andrewoid) [inline src/main/java/smartsnake/neuralnetwork/runge/LearnAndSave.java:37]: Seems like a strange comment.

**Comment by emmyrunge** [comment]
https://github.com/andrewoid/SmartSnake/assets/96812396/58431173-ecf1-48c6-950a-4fcecfeb067d

**Review by andrewoid — APPROVED** [review]

**Pull request merged**